bsr_dense_bmm(): enable more precise float32 support with float64 accumulators

nikitaved opened the pull request:
Float64 is there in Triton! This PR increases precision for float32 inputs with float64 accumulation dtype.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101163
* __->__ #100882

nikitaved committed:
bsr_dense_bmm(): enable float32 support with float64 accumulators

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_bmm(): enable more precise float32 support with float64 accumulators"

Float64 is there in Triton! This PR increases precision for float32 inputs with float64 accumulation dtype.




[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_bmm(): enable more precise float32 support with float64 accumulators"

Float64 is there in Triton! This PR increases precision for float32 inputs with float64 accumulation dtype.




[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_bmm(): enable more precise float32 support with float64 accumulators"

Float64 is there in Triton! This PR increases precision for float32 inputs with float64 accumulation dtype.




[ghstack-poisoned]

cpuhrsch commented:
The failure unfortunately looks related. It's surprising because the accumulation precision here is higher and thus we'd expect the results to be closer not further away. We might need to adjust tolerances here.

nikitaved commented:
The result is very stange. It has the same error as with `float32` accumulators...

nikitaved committed:
Update on "bsr_dense_bmm(): enable more precise float32 support with float64 accumulators"

Float64 is there in Triton! This PR increases precision for float32 inputs with float64 accumulation dtype.




[ghstack-poisoned]

amjames commented on torch/sparse/_triton_ops.py:
```diff
@@ -195,7 +196,7 @@ def _bsr_strided_dense_rowspace_kernel(
+        output_acc_block = tl.zeros((BLOCKSIZE_ROW, BLOCKSIZE_ROW), dtype=acc_dtype)
         for _ in range(row_nnz):
             values_block = tl.load(values_block_ptrs)
 
```
It won't let me add a suggestion outside of the diff, but playing with this locally the tests pass if you modify the `tl.dot` call found a few lines below:

```python
output_acc_block += tl.dot(values_block, dense_block, allow_tf32=False)
```

You can make toggling the flag conditional on the dtype in play, but as far as I know it should only have an effect for float32. 

Note that things do not quite work out if we set that flag to match the value of `torch.backends.cuda.matmul.allow_tf32`. When torch has this flag set true and the triton flag is also set true the results do not match, and the errors are too large for adjusting the dtype to account for them (max ~10^-1 absolute error).

nikitaved replied:
Thank you! I will play with it. I wonder whether it affects all mm-like methods...

nikitaved committed:
Update on "bsr_dense_bmm(): enable more precise float32 support with float64 accumulators"

Float64 is there in Triton! This PR increases precision for float32 inputs with float64 accumulation dtype.




[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_bmm(): enable more precise float32 support with float64 accumulators"

Float64 is there in Triton! This PR increases precision for float32 inputs with float64 accumulation dtype.




[ghstack-poisoned]

nikitaved commented:
@pytorchbot merge
